enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,940 @@
+
+### Version change support
+
+- Because we want to maintain upgrade expectations with Kubernetes and OpenShift,
+  we will only support changing versions one Y version at a time (x.y to x.y+1).
+- Because we cannot guarantee that the data formats between Y versions are
+  compatible after an upgrade, we will only support rolling back to a previous
+  Y version when restoring from a backup.
+- Because we may need to support "manual" changes to correct for regressions
+  within a Y version, and because we expect the storage format and other data
+  to be forward and backward compatible within a Y version, we will support
+  changing from any Z version to any other Z version for the same version of Y,
+  including downgrading.
+- Because we may need to block certain upgrade sequences, similar to OpenShift's
+  upgrade graph, but we cannot ensure access to that upgrade graph from edge
+  systems and we cannot prevent an attempted upgrade via a new ostree deployment,
+  we will incorporate a mechanism to block specific upgrades by listing version
+  numbers _from which_ a new version cannot be upgraded (X.Y+1.Z may include
+  X.Y.Z in its "block" list). When a new version detects that the system is
+  upgrading from a version in that block list, it will refuse to start and
+  cause a greenboot failure.
+
+### Backup retention
+
+- Because a user may stage multiple ostree deployments on a host and boot them in
+  any order, we will keep multiple backups to ensure that we can roll back to a
+  state compatible with any ostree deployment
+- Because we want to minimize the impact of backups on storage requirements,
+  we will keep only 1 backup per ostree deployment.
+
+### Backup format
+
+- Because we want the backup process to be simple and we want to minimize the
+  space used by each backup, we will use `cp` with reflinks to create copy-on-write
+  versions of all of the content being backed up.
+
+### Kubernetes storage format upgrades
+
+- Because we need to support API version deprecation and removal in Kubernetes
+  and OpenShift APIs, we will run the storage version migrator to update all
+  data in the database when a Y version update is detected.
+
+### Order of actions
+
+- To ensure backup and restore process does not result in data corruption, it
+  will be performed with MicroShift cluster not running.
+- To allow database upgrades only etcd and kube-apiserver will be active during
+  data migration.
+- System start was chosen as point in time that above actions will happen,
+  just before starting MicroShift cluster.
+
+### Workflow Description
+
+**MicroShift administrator** is a human responsible for preparing
+ostree commits and scheduling devices to use these commits.
+
+Upgrade:
+
+1. MicroShift administrator prepares a new ostree commit
+1. MicroShift administrator schedules device to reboot and use new ostree commit
+1. Device boots new commit
+1. Operating System, greenboot, and MicroShift take actions (migrating database
```

Comment:
because microshift ships its own "special" etcd, we consider etcd to be part of it and implicitly included in backup/restore/migration

Reply:
i don't think this answers my question

Reply:
Please take a look at https://github.com/openshift/enhancements/pull/1387#discussion_r1192024681
Does it answer your question?
I pushed new commit to use "database" only in relation to etcd, otherwise it's "data" (MicroShift's as a whole)

Reply:
yes, thanks